hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/scheduler/placement/policy/MultiComparatorPolicy.java

```diff
@@ -0,0 +1,346 @@
+ * </p>
+ */
+public class MultiComparatorPolicy<N extends SchedulerNode>
+    implements MultiNodeLookupPolicy<N>, Configurable {
+
+  private static final Logger LOG =
+      LoggerFactory.getLogger(MultiComparatorPolicy.class);
+  // comparators
+  private static final DominantResourceCalculator DOMINANT_RC =
+      new DominantResourceCalculator();
+  private static final Map<ComparatorKey, Function<SchedulerNode, Comparable>>
+      COMPARATOR_CALCULATORS = Collections.unmodifiableMap(
+      new HashMap<ComparatorKey, Function<SchedulerNode, Comparable>>() {{
+        // for vcores
+        put(ComparatorKey.ALLOCATED_VCORES,
+            obj -> obj.getAllocatedResource().getVirtualCores());
+        put(ComparatorKey.UNALLOCATED_VCORES,
+            obj -> obj.getUnallocatedResource().getVirtualCores());
+        put(ComparatorKey.TOTAL_VCORES,
+            obj -> obj.getTotalResource().getVirtualCores());
+        // for memory
+        put(ComparatorKey.ALLOCATED_MEMORY,
+            obj -> obj.getAllocatedResource().getMemorySize());
+        put(ComparatorKey.UNALLOCATED_MEMORY,
+            obj -> obj.getUnallocatedResource().getMemorySize());
+        put(ComparatorKey.TOTAL_MEMORY,
+            obj -> obj.getTotalResource().getMemorySize());
+        // for resource
+        put(ComparatorKey.ALLOCATED_RESOURCE,
+            SchedulerNode::getAllocatedResource);
+        put(ComparatorKey.UNALLOCATED_RESOURCE,
+            SchedulerNode::getUnallocatedResource);
+        put(ComparatorKey.TOTAL_RESOURCE,
+            SchedulerNode::getTotalResource);
+        // for dominant ratio
+        put(ComparatorKey.DOMINANT_ALLOCATED_RATIO, obj -> Resources
+            .ratio(DOMINANT_RC, obj.getAllocatedResource(),
+                obj.getTotalResource()));
+        // for node ID
+        put(ComparatorKey.NODE_ID, SchedulerNode::getNodeID);
+      }});
+  // conf keys and default values
+  public static final String COMPARATORS_CONF_KEY = "comparators";
+  protected static final List<Comparator> DEFAULT_COMPARATORS = Collections
+      .unmodifiableList(Arrays.asList(
+          new Comparator(ComparatorKey.DOMINANT_ALLOCATED_RATIO,
+              OrderDirection.ASC, COMPARATOR_CALCULATORS
+              .get(ComparatorKey.DOMINANT_ALLOCATED_RATIO)),
+          new Comparator(ComparatorKey.NODE_ID, OrderDirection.ASC,
+              COMPARATOR_CALCULATORS.get(ComparatorKey.NODE_ID))));
+
+  protected Map<String, Set<N>> nodesPerPartition = new ConcurrentHashMap<>();
+  protected List<Comparator> comparators;
+  private Configuration conf;
+
+  public MultiComparatorPolicy() {
+  }
+
+  @Override
+  public void setConf(Configuration conf) {
+    // init comparators
+    this.comparators = DEFAULT_COMPARATORS;
+    if (conf == null) {
+      return;
+    }
+    this.conf = conf;
+    String policyName = conf.get(
+        CapacitySchedulerConfiguration.MULTI_NODE_SORTING_POLICY_CURRENT_NAME);
```

Comment:
what is the difference between `MULTI_NODE_SORTING_POLICY_CURRENT_NAME` and ``MULTI_NODE_SORTING_POLICY`

Reply:
Currently`MULTI_NODE_SORTING_POLICY_NAME` is the prefix of all the configurations of multi-node policy: `yarn.scheduler.capacity.multi-node-sorting.policy`, it's not a proper name, I would like rename it as MULTI_NODE_SORTING_POLICY in next update. `MULTI_NODE_SORTING_POLICY_CURRENT_NAME` is used to transfer the policyName to policy instance.